aten/src/ATen/native/ForeachUtils.h

```diff
@@ -173,5 +179,79 @@ bool can_use_fast_route(TensorList tensors1, TensorList tensors2, bool does_op_p
   return can_use_fast_route({tensors1, tensors2}, {}, does_op_promote_integer_inputs_to_float);
 }
 
+
+using DeviceDtypeKey = std::pair<at::Device, at::ScalarType>;
+using IndicesT = std::vector<int>;
+using nested_optional_tensorvec_t = std::vector<std::vector<c10::optional<at::Tensor>>>;
+using TensorsAndIndicesT = std::pair<nested_optional_tensorvec_t, IndicesT>;
+using FlatMap = std::unordered_map<DeviceDtypeKey, TensorsAndIndicesT, ParamsHash<DeviceDtypeKey>>;
+
+FlatMap group_tensors_by_first_tensors_device_and_dtype(const nested_optional_tensorvec_t& nested_tensorlist, const bool with_indices) {
```

Comment:
Should we have this be private? Partly because we don't intend for users to use this explicitly and also partly to give it some bake in time in case anything needs to change about it.

Reply:
it sounds about right but how would users use this function? In their C++/CUDA extensions?

Reply:
Yea. I'd lean towards making things private anyway because, in my experience, people sometimes uses our private APIs when we don't expect them to and cause pain for us in the future when modifying them. I'd like to avoid that as much as possible.

Reply:
I get your point. I know Python APIs are kind of guaranteed. Though from my perspective, ATen functions have seemed to be allowed to change drastically so that sometimes extensions fail to compile.